File: IsomorphicStrings.java
class Solution {
    public boolean isIsomorphic(String s, String t) {
        if(s.length() != t.length()){
            return false;
        }
        
        HashMap<Character,Character> characterMap = new HashMap<>();
        HashMap<Character, Boolean> uniqueMap = new HashMap<>();
        
        for(int i=0; i<s.length(); i++){
            char ch1 = s.charAt(i);
            char ch2 = t.charAt(i);
            
            if(characterMap.containsKey(ch1) == true){
                if(characterMap.get(ch1) != ch2){
                    return false;
                }
            }else{
                if(uniqueMap.containsKey(ch2) == true){
                        return false;
                    }else{
                        characterMap.put(ch1,ch2);
                        uniqueMap.put(ch2, true);
                    }
                }
            }
            
        
        
        return true;
        
    }
}

File: IsomorphicStrings.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
// TC: O(N) SC: O(N)
var isIsomorphic = function(s, t) {
    if(s.length !== t.length) return false;
    const map = new Map();
    const set = new Set();
    for(let i=0; i<s.length; i++) {
        if(map.has(s[i]) && map.get(s[i]) !== t[i]) return false;      
        if(!map.has(s[i]) && set.has(t[i])) return false;
        map.set(s[i], t[i]);
        set.add(t[i]);
    }
    return true;
};
